azure/Guidelines.md

```diff
@@ -0,0 +1,864 @@
+
+#### Exactly Once Behavior = Client Retries & Service Idempotency
+> :white_check_mark: **DO** ensure that __all__ HTTP methods are idempotent.
+ 
+> :ballot_box_with_check: **YOU SHOULD** use PUT or PATCH to create a resource as these HTTP methods are easy to implement, allow the customer to name their own resource, and are idempotent.
+ 
+> :heavy_check_mark: **YOU MAY** use POST to create a resource but you must make it idempotent and, of course, the response __must__ return the URL of the created resource with a 201-Created. One way to make POST idempotent is to use the Repeatability-Request-ID & Repeatability-First-Sent headers (See [Repeatability of requests](#Repeatability-of-requests)).
+
+##### HTTP Return Codes
+> :white_check_mark: **DO** adhere to the return codes in the following table when the method completes synchronously and is successful:
+
+Method | Description | Response Status Code 
+-------|-------------|---------------------
+PATCH  | Create/Modify the resource with JSON Merge Patch | 200-OK, 201-Created
+PUT    | Create/Replace the *whole* resource | 200-OK, 201-Created 
+POST   | Create new resource (ID set by service) | 201-Created with URL of created resource
+GET    | Read (i.e. list) a resource collection | 200-OK
+GET    | Read the resource | 200-OK 
+DELETE | Remove the resource | 204-No Content\; avoid 404-Not Found 
+
+> :white_check_mark: **DO** return status code ```202-Accepted``` and follow the guidance in [Long Running Operations & Jobs](#long-running-operations--jobs) when a PUT, POST, or DELETE method completes asynchronously
+
+> :white_check_mark: **DO** treat method names as case sensitive and should always be in uppercase
+> 
+> :white_check_mark: **DO** return the state of the resource after a PUT, PATCH, POST, or GET operation with a ```200-OK``` or ```201-Created```.
+> 
+> :white_check_mark: **DO** return a ```204-No Content``` without a resource/body for a DELETE operation (even if the URL identifies a resource that does not exist; do not return ```404-Not Found```)
+> 
+> :white_check_mark: **DO** support caching and optimistic concurrency by honoring the the if-match, if-none-match, if-modified-since, and if-unmodified-since request headers and by returning the etag and last-modified response headers
+
+### HTTP Query Parameters and Header Values
+Because information in the service URL, as well as the request / response, are strings, there must be a predictable, well-defined scheme to convert strings to their corresponding values.
+
+> :white_check_mark: **DO** validate all query parameter and request header values and fail the operation with ```400-Bad Request``` if any value fails validation. Return an error response as described in the [Handling Errors](#Handling-errors) section indicating what is wrong so customer can diagnose the issue and fix it themselves.
+> 
+> :white_check_mark: **DO** use the following table when translating strings:
```

Comment:
It seems like Azure services are often implemented in .NET. I wonder if we should not provide a .NET library that would help service teams implement some of the behaviors recommended by this document. Possibly some of these APIs already exist in .NET, and in such case, we would just provide a sample/guide how to use them, not a library.

Reply:
Great idea!